concepts/pointers/about.md

```diff
@@ -0,0 +1,404 @@
+
+Yes we can!
+We can pass in the addresses of `a` and `b` as pointers.
+But, what good will that do, since the function copies the pointers?
+It is good, because a copied address is still the same address.
+By dereferencing the address we can get access to the original value as it was defined in `main`
+
+```c
+#include <stdio.h>
+
+// pass in pointers to a and b
+void swap(int *a, int *b) {
+    // set temp to value at a's dereferenced address
+    int temp = *a;
+
+    // set a's value to value at b's dereferenced address
+    *a = *b;
+
+    // set b's value to temp
+    *b = temp;
+
+    // prints: In 'swap', a is 4 and b is 3.
+    printf("In 'swap', a is %d and b is %d.\n", *a, *b);
+}
+
+int main() {
+    int a = 3, b = 4;
+    swap(&a, &b);
+    // prints: After 'swap', a is 4 and b is 3.
+    printf("After 'swap', a is %d and b is %d.\n", a, b);
+    return 0;
+}
+```
+
+That's a lot of `*`s in `swap` now, but it now works as intended.
+
+A way to read the dereference operator is "value at".
+So, `*a` is the value at `a`.
+How well does this work with strings?
+
+## Pointers to Char
+
+We know that the value of an array variable is the address where the array starts.
+And we know we can define a string as either a char array, `char my_string[]`, or as a pointer to `char`, `char *`.
+How are they different?
+
+In this code we pass a `char *` argument to `describe_me`.
+The `adjective` argument is defined as a `char` array and is accepted by the function as a pointer to `char`.
+
+```c
+#include <stdio.h>
+#include <string.h>
+
+// takes pointer to char
+void describe_me(char *adjective){
+    strcpy(adjective, "evil");
+    // prints: I am evil.
+    printf("I am %s.\n", adjective);
+}
+
+int main() {
+    // defined as a char array
+    char adjective[] = "good";
+    describe_me(adjective);
+    // prints: I am really evil.
+    printf("I am really %s.\n", adjective);
+    
+    return 0;
+}
+```
+
+We are able to make a persistent change to `adjective`, demonstrating that the defined char array can be accepted as a pointer to char.
+
+Now let's try defining `adjective` as a pointer to `char` directly.
+
+```c
+// Online C compiler to run C program online
+#include <stdio.h>
+#include <string.h>
+
+void describe_me(char *adjective){
+    // segmentation fault on this line
+    strcpy(adjective, "evil");
+    printf("I am %s.\n", adjective);
+}
+
+int main() {
+    char *adjective= "good";
+    describe_me(adjective);
+    printf("I am really %s.\n", adjective);
+    
+    return 0;
+}
+```
+
+Remember how pointers are powerful, but also potentially dangerous?
+But why doesn't it work?
+An array is a pointer and a pointer worked with a char array!
+The reason is that, before, `adjective` was defined as a `char` array which is a variable belonging to `main`.
+Now, `adjective` is defined as a string _literal_, meaning that it is written to read-only static memory which belongs to the entire program.
```

Comment:
Not exactly correct wording here I feel.
`adjective` is still just a pointer, the same as any other pointer.
What it _points to_ is a string literal.

Reply:
Yes, that's technically correct. I elided talk about pointers to focus on the difference between local char arrays and string literals. Rather than do something like 

```
The reason is that, before, `adjective` was defined as a pointer to a `char` array which is a variable belonging to `main`.
Now, `adjective` is defined as a pointer to a string _literal_, meaning that it is written to read-only static memory which belongs to the entire program.
```

I'm thinking to do something like

```
The reason is that, before, `adjective` was defined as a `char` array which is a variable belonging to `main`.
Now, `adjective` is defined as a string _literal_, meaning that it is written to read-only static memory which belongs to the entire program. They are both pointers, but they point to different things.
```

But I suppose the first way isn't so bad, but, then again, maybe it's not so bad because I already understand it. For a novice, I'm not sure which way would be clearer to assimilate.